akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala

```diff
@@ -54,212 +60,86 @@ private[http2] trait Http2MultiplexerSupport { logic: GraphStageLogic with Stage
-          debug(s"Updating window for $streamId by $increment to ${windowLeftFor(streamId)} buffered bytes: ${streamFor(streamId).bufferedBytes}")
-        }
-
-      override def cancelSubStream(streamId: Int): Unit = streamFor(streamId).cancelStream()
       override def updateMaxFrameSize(newMaxFrameSize: Int): Unit = currentMaxFrameSize = newMaxFrameSize
       override def updateDefaultWindow(newDefaultWindow: Int): Unit = {
-        val delta = newDefaultWindow - currentInitialWindow
+        val delta = newDefaultWindow - _currentInitialWindow
+        _currentInitialWindow = newDefaultWindow
 
-        currentInitialWindow = newDefaultWindow
-        outStreams.values.foreach(i => updateWindowFor(i.streamId, delta))
+        distributeWindowDeltaToAllStreams(delta)
       }
       override def updatePriority(info: PriorityFrame): Unit = prioritizer.updatePriority(info)
 
-      private def streamFor(streamId: Int): OutStream = outStreams.get(streamId) match {
-        case None =>
-          val newOne = new OutStream(streamId, None, currentInitialWindow)
-          outStreams += streamId -> newOne
-          newOne
-        case Some(old) => old
-      }
-      private def windowLeftFor(streamId: Int): Int = streamFor(streamId).outboundWindowLeft
-      private def updateWindowFor(streamId: Int, increment: Int): Unit = {
-        val info = streamFor(streamId)
-        info.outboundWindowLeft += increment
-        if (info.canSend) enqueueOutStream(info)
-      }
+      def enqueueOutStream(streamId: Int): Unit = updateState(_.enqueueOutStream(streamId))
+      def closeStream(streamId: Int): Unit = updateState(_.closeStream(streamId))
 
-      def enqueueOutStream(outStream: OutStream): Unit = state.enqueueOutStream(outStream)
-      def closeStream(outStream: OutStream): Unit = state.closeStream(outStream)
+      def onPull(): Unit = updateState(_.onPull())
 
-      override def shutdown(): Unit = outStreams.values.foreach(_.cancelStream())
+      private var _state: MultiplexerState = Idle
 
-      var state: MultiplexerState = Idle
-      def onPull(): Unit = state.onPull()
-      private def become(nextState: MultiplexerState): Unit = {
-        if (nextState.name != state.name) recordStateChange(state.name, nextState.name)
+      private def updateState(transition: MultiplexerState => MultiplexerState): Unit = {
+        val oldState = _state
+        val newState = transition(_state)
+        _state = newState
 
-        state = nextState
+        if (newState.name != oldState.name) recordStateChange(oldState.name, newState.name)
       }
 
       private[http2] sealed trait MultiplexerState extends Product {
         def name: String = productPrefix
 
-        def onPull(): Unit
-        def pushControlFrame(frame: FrameEvent): Unit
-        def connectionWindowAvailable(): Unit
-        def enqueueOutStream(outStream: OutStream): Unit
-        def closeStream(outStream: OutStream): Unit
+        def onPull(): MultiplexerState
```

Comment:
The one significant change in multiplexer is returning states instead of relying on side-effects for state changes.